File: javascript/min_max.js
var max = function (a, b) {
  return a > b ? a : b;
};

var min = function (a, b) {
  return a > b ? b : a;
};

var minmax = function (a, b) {
  return a > b ? {min: b, max: a} : {min: a, max: b};
};

var max_element = function (ary) {
  return Array.prototype.reduce.call(ary, max);
};

var min_element = function (ary) {
  return Array.prototype.reduce.call(ary, min);
};

var minmax_element = function (ary) {
  return {
    min_element: min_element(ary),
    max_element: max_element(ary)
  };
};

var lexicographical_compare = function (ary1, ary2) {
  if (ary2.length < ary1.length)
    return false;
  for (var i = 0; i < ary1.length; i++) {
    if (ary1[i] < ary2[i]) {
      return true;
    }
    else if (ary2[i] < ary1[i]) {
      return false;
    }
  }
  return true;
};

var is_permutation = function (ary1, ary2) {
  if (ary1.length != ary2.length) return false;
  elements = {};
  Array.prototype.forEach.call(ary1, function(e) {
    if (elements[e]) {
      elements[e] += 1;
    }
    else {
      elements[e] = 1;
    }
  });
  for (var i = 0; i < ary2.length; i++) {
    if (elements[ary2[i]] && elements[ary2[i]] > 0) {
      elements[ary2[i]]--;
    }
    else {
      return false;
    }
  }
  return true;
};

var next_permutation = function (ary) {
  if (ary.length <= 1) return false;
  var pos1, pos2, tmp, part;
  var i = ary.length - 1;
  while (true) {
    pos1 = i--;
    if (ary[i] < ary[pos1]) {
      pos2 = ary.length - 1;
      while (ary[i] >= ary[pos2])
        pos2--;
      tmp = ary[i];
      ary[i] = ary[pos2];
      ary[pos2] = tmp;
      var pivot = (ary.length - pos1) / 2;
      for (var j = 0; j < pivot; j++) {
        tmp = ary[ary.length - j - 1];
        ary[ary.length - j - 1] = ary[pos1 + j];
        ary[pos1 + j] = tmp;
      }
      return true;
    }
    if (i === 0) {
      Array.prototype.reverse.call(ary);
      return false;
    }
  }
};

var prev_permutation = function (ary) {
  if (ary.length <= 1) return false;
  var pos1, pos2, tmp, part;
  var i = ary.length - 1;
  while (true) {
    pos1 = i--;
    if (ary[pos1] < ary[i]) {
      pos2 = ary.length - 1;
      while (ary[pos2] >= ary[i])
        pos2--;
      tmp = ary[i];
      ary[i] = ary[pos2];
      ary[pos2] = tmp;
      var pivot = (ary.length - pos1) / 2;
      for (var j = 0; j < pivot; j++) {
        tmp = ary[ary.length - j - 1];
        ary[ary.length - j - 1] = ary[pos1 + j];
        ary[pos1 + j] = tmp;
      }
      return true;
    }
    if (i === 0) {
      Array.prototype.reverse.call(ary);
      return false;
    }
  }
};

module.exports = {
  max: max,
  min: min,
  max_element: max_element,
  min_element: min_element,
  minmax: minmax,
  minmax_element: minmax_element,
  lexicographical_compare: lexicographical_compare,
  is_permutation: is_permutation,
  next_permutation: next_permutation,
  prev_permutation: prev_permutation
};

File: rust/src/min_max.rs
pub fn max<T: PartialOrd>(a: T, b: T) -> T {
    if a < b {
        b
    } else {
        a
    }
}

pub fn min<T: PartialOrd>(a: T, b: T) -> T {
    if a < b {
        a
    } else {
        b
    }
}

pub fn max_element<T: PartialOrd>(iterable: impl IntoIterator<Item = T>) -> Option<T> {
    let iter = iterable.into_iter();
    iter.reduce(|a, b| if a < b { b } else { a })
}

pub fn min_element<T: PartialOrd>(iterable: impl IntoIterator<Item = T>) -> Option<T> {
    let iter = iterable.into_iter();
    iter.reduce(|a, b| if a < b { a } else { b })
}

pub fn min_max<T: PartialOrd>(a: T, b: T) -> (T, T) {
    if a < b {
        (a, b)
    } else {
        (b, a)
    }
}

pub fn min_max_element<T: PartialOrd + Clone>(
    iterable: impl IntoIterator<Item = T>,
) -> Option<(T, T)> {
    let mut iter = iterable.into_iter();
    let first = iter.next()?;
    let (mut min_val, mut max_val) = (first.clone(), first);
    for item in iter {
        if item < min_val {
            min_val = item.clone();
        }
        if item > max_val {
            max_val = item;
        }
    }
    Some((min_val, max_val))
}

pub fn lexicographical_compare<T: PartialOrd>(
    a: impl IntoIterator<Item = T>,
    b: impl IntoIterator<Item = T>,
) -> bool {
    let mut iter_a = a.into_iter();
    let mut iter_b = b.into_iter();

    loop {
        match (iter_a.next(), iter_b.next()) {
            (Some(x), Some(y)) => {
                if x < y {
                    return true;
                } else if x > y {
                    return false;
                }
            }
            (None, Some(_)) => return true,
            (Some(_), None) => return false,
            (None, None) => return false, // not <
        }
    }
}

pub fn is_permutation<T: PartialOrd + Clone>(
    a: impl IntoIterator<Item = T>,
    b: impl IntoIterator<Item = T>,
) -> bool {
    let mut vec_a: Vec<T> = a.into_iter().collect();
    let mut vec_b: Vec<T> = b.into_iter().collect();

    if vec_a.len() != vec_b.len() {
        return false;
    }

    vec_a.sort_by(|x, y| x.partial_cmp(y).unwrap());
    vec_b.sort_by(|x, y| x.partial_cmp(y).unwrap());

    vec_a == vec_b
}

pub fn next_permutation<T: PartialOrd + Clone>(data: &mut [T]) -> bool {
    let n = data.len();
    if n < 2 {
        return false;
    }

    let mut k = None;
    for i in (0..n - 1).rev() {
        if data[i] < data[i + 1] {
            k = Some(i);
            break;
        }
    }
    let k = match k {
        Some(index) => index,
        None => {
            data.reverse();
            return false;
        }
    };

    let mut l = None;
    for i in (k + 1..n).rev() {
        if data[k] < data[i] {
            l = Some(i);
            break;
        }
    }
    let l = l.unwrap();

    data.swap(k, l);

    data[k + 1..].reverse();

    true
}

pub fn prev_permutation<T: PartialOrd + Clone>(data: &mut [T]) -> bool {
    let n = data.len();
    if n < 2 {
        return false;
    }

    let mut k = None;
    for i in (0..n - 1).rev() {
        if data[i] > data[i + 1] {
            k = Some(i);
            break;
        }
    }
    let k = match k {
        Some(index) => index,
        None => {
            data.reverse();
            return false;
        }
    };

    let mut l = None;
    for i in (k + 1..n).rev() {
        if data[k] > data[i] {
            l = Some(i);
            break;
        }
    }
    let l = l.unwrap();

    data.swap(k, l);

    data[k + 1..].reverse();

    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_max() {
        assert_eq!(max(3, 5), 5);
        assert_eq!(max(10, -2), 10);
        assert_eq!(max(0.5, 0.3), 0.5);
        assert_eq!(max('a', 'b'), 'b');
    }

    #[test]
    fn test_min() {
        assert_eq!(min(3, 5), 3);
        assert_eq!(min(10, -2), -2);
        assert_eq!(min(0.5, 0.3), 0.3);
        assert_eq!(min('a', 'b'), 'a');
    }

    #[test]
    fn test_max_element() {
        let numbers = vec![3, 1, 4, 1, 5, 9];
        assert_eq!(max_element(numbers), Some(9));

        let empty: Vec<i32> = vec![];
        assert_eq!(max_element(empty), None);

        let chars = vec!['a', 'c', 'b'];
        assert_eq!(max_element(chars), Some('c'));
    }

    #[test]
    fn test_min_element() {
        let numbers = vec![3, 1, 4, 1, 5, 9];
        assert_eq!(min_element(numbers), Some(1));

        let empty: Vec<i32> = vec![];
        assert_eq!(min_element(empty), None);

        let chars = vec!['a', 'c', 'b'];
        assert_eq!(min_element(chars), Some('a'));
    }

    #[test]
    fn test_min_max() {
        assert_eq!(min_max(3, 5), (3, 5));
        assert_eq!(min_max(10, -2), (-2, 10));
        assert_eq!(min_max(0.5, 0.3), (0.3, 0.5));
        assert_eq!(min_max('a', 'b'), ('a', 'b'));
    }

    #[test]
    fn test_min_max_element() {
        let numbers = vec![3, 1, 4, 1, 5, 9];
        assert_eq!(min_max_element(numbers), Some((1, 9)));

        let empty: Vec<i32> = vec![];
        assert_eq!(min_max_element(empty), None);

        let chars = vec!['a', 'c', 'b'];
        assert_eq!(min_max_element(chars), Some(('a', 'c')));
    }

    #[test]
    fn test_lexicographical_compare() {
        let a = "abc";
        let b = "abd";
        assert!(lexicographical_compare(a.chars(), b.chars()));
        assert!(!lexicographical_compare(b.chars(), a.chars()));
        assert!(!lexicographical_compare(a.chars(), a.chars()));

        let x = vec![1, 2, 3];
        let y = vec![1, 2, 4];
        assert!(lexicographical_compare(x, y));

        let x = vec![1, 2, 3, 4];
        let y = vec![1, 2, 3];
        assert!(!lexicographical_compare(x, y));
    }

    #[test]
    fn test_is_permutation() {
        let a = vec![1, 2, 3];
        let b = vec![3, 2, 1];
        assert!(is_permutation(a, b));

        let c = vec![1, 2, 2];
        let d = vec![2, 1, 1];
        assert!(!is_permutation(c, d));

        let e = vec!['a', 'b', 'c'];
        let f = vec!['c', 'b', 'a'];
        assert!(is_permutation(e, f));

        let g = vec![1, 2, 3];
        let h = vec![1, 2, 3, 4];
        assert!(!is_permutation(g, h));
    }

    #[test]
    fn test_next_permutation() {
        let mut data = vec![1, 2, 3];
        assert!(next_permutation(&mut data));
        assert_eq!(data, vec![1, 3, 2]);

        assert!(next_permutation(&mut data));
        assert_eq!(data, vec![2, 1, 3]);

        assert!(next_permutation(&mut data));
        assert_eq!(data, vec![2, 3, 1]);

        assert!(next_permutation(&mut data));
        assert_eq!(data, vec![3, 1, 2]);

        assert!(next_permutation(&mut data));
        assert_eq!(data, vec![3, 2, 1]);

        assert!(!next_permutation(&mut data));
        assert_eq!(data, vec![1, 2, 3]);
    }

    #[test]
    fn test_prev_permutation() {
        let mut data = vec![3, 2, 1];
        assert!(prev_permutation(&mut data));
        assert_eq!(data, vec![3, 1, 2]);

        assert!(prev_permutation(&mut data));
        assert_eq!(data, vec![2, 3, 1]);

        assert!(prev_permutation(&mut data));
        assert_eq!(data, vec![2, 1, 3]);

        assert!(prev_permutation(&mut data));
        assert_eq!(data, vec![1, 3, 2]);

        assert!(prev_permutation(&mut data));
        assert_eq!(data, vec![1, 2, 3]);

        assert!(!prev_permutation(&mut data));
        assert_eq!(data, vec![3, 2, 1]);
    }
}
